### PR TITLE
[SYCL-MLIR] Represent all SYCLMethodOpInterface arguments in SSA form

### PR DESCRIPTION
**Pull request by victor-eds**
Do not use memref types in the arguments of operations implementing this interface.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td [hunk 1]
```diff
@@ -548,7 +548,7 @@ def SYCLCallOp : SYCL_Op<"call", [CallOpInterface]> {
 // accessor.subscript OPERATION
 ////////////////////////////////////////////////////////////////////////////////
 
-def SYCLAccessorSubscriptIndex : AnyTypeOf<[IndexType, IDMemRef]>;
+def SYCLAccessorSubscriptIndex : AnyTypeOf<[IndexType, SYCL_IDType]>;
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:551]: Only signature change.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpInterfaces.td [hunk 2]
```diff
@@ -43,7 +43,23 @@ def SYCLMethodOpInterface : SYCL_OpInterface<"SYCLMethodOpInterface"> {
     }], "llvm::ArrayRef<llvm::StringLiteral>", "getMethodNames">,
     InterfaceMethod<[{
       Return the SYCL type this method is a member of.
-    }], "::mlir::Type", "getBaseType">,
+    }], "::mlir::Type", "getBaseType", /*args=*/ (ins),
+    /*methodBody=*/ [{
+      return $_op.getArgumentTypes()[0].template cast<::mlir::TypeAttr>().getValue();
+    }]
+    >,
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpInterfaces.td:50]: We keep this function and its semantics.

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpInterfaces.td [hunk 3]
```diff
@@ -43,7 +43,23 @@ def SYCLMethodOpInterface : SYCL_OpInterface<"SYCLMethodOpInterface"> {
...
+    }], "::mlir::SmallVector<::mlir::Type>", "getArgumentTypes",
+    /*args=*/ (ins), /*methodBody=*/ [{
+      SmallVector<Type> Res;
+      const auto Tys =
+        $_op.getArgumentTypes().template getAsRange<::mlir::TypeAttr>();
+      std::transform(Tys.begin(), Tys.end(), std::back_inserter(Res),
+                     [](auto Attr) { return Attr.getValue(); });
+      return Res;
+    }]
+    >,
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOpInterfaces.td:62]: While also adding this new one.

File: mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares helper functions to generate SYCL method operations.
+//
+//===----------------------------------------------------------------------===//
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h:11]: Utility header.

File: mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+namespace mlir {
+class Location;
+class OpBuilder;
+class ValueRange;
+namespace sycl {
+class SYCLMethodOpInterface;
+/// Generates a list of values to be used as arguments to a
+/// SYCLMethodOpInterface instance from \p Original.
+SmallVector<Value> adaptSYCLMethodOpArguments(OpBuilder &Builder, Location Loc,
+                                              ValueRange Original);
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h:29]: Functionality migrated from cgeist and extended to support a range of values.

File: mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+/// Abstracts different cast operations from which \p Original may have
+/// originated.
+Value abstractCasts(Value Original);
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/MethodUtils.h:32]: Migrated from cgeist.

File: mlir-sycl/lib/Dialect/IR/CMakeLists.txt [hunk 7]
```diff
@@ -22,6 +23,7 @@ else()
     MLIRVectorInterfaces
     MLIRLLVMDialect
     MLIRMemRefDialect
+    MLIRPolygeist
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/IR/CMakeLists.txt:26]: No circular dependency.

File: mlir-sycl/lib/Dialect/IR/SYCLOps.cpp [hunk 8]
```diff
@@ -119,10 +119,7 @@ LogicalResult SYCLAccessorSubscriptOp::verify() {
   };
 
   return TypeSwitch<Type, LogicalResult>(getOperand(1).getType())
-      .Case<MemRefType>([&](auto MemRefTy) {
-        Type ElemTy = MemRefTy.getElementType();
-        auto IDTy = ElemTy.dyn_cast<IDType>();
-        assert(IDTy && "Unhandled input memref type");
+      .Case<IDType>([&](auto IDTy) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/IR/SYCLOps.cpp:122]: Verification has to change too.

File: mlir-sycl/test/Transforms/sycl-method-to-sycl-call.mlir [hunk 9]
```diff
@@ -10,19 +10,22 @@
...
-// CHECK-NEXT:      %[[VAL_7:.*]] = sycl.call(%[[VAL_6]], %[[VAL_1]]) {FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIiLi2ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi2EvEERiNS0_2idILi2EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_2_i32_rw_gb, 4>, memref<?x!sycl_id_2_>) -> memref<?xi32, 4>
+// CHECK-NEXT:      memref.store %[[VAL_1]], %[[ID_ALLOCA]]{{\[}}%[[VAL_2]]] : memref<1x!sycl_id_2_>
+// CHECK-NEXT:      %[[ID_CAST:.*]] = memref.cast %[[ID_ALLOCA]] : memref<1x!sycl_id_2_> to memref<?x!sycl_id_2_>
+// CHECK-NEXT:      %[[VAL_7:.*]] = sycl.call(%[[VAL_6]], %[[ID_CAST]]) {FunctionName = @"operator[]", MangledFunctionName = @_ZNK4sycl3_V18accessorIiLi2ELNS0_6access4modeE1026ELNS2_6targetE2014ELNS2_11placeholderE0ENS0_3ext6oneapi22accessor_property_listIJEEEEixILi2EvEERiNS0_2idILi2EEE, TypeName = @accessor} : (memref<?x!sycl_accessor_2_i32_rw_gb, 4>, memref<?x!sycl_id_2_>) -> memref<?xi32, 4>
 // CHECK-NEXT:      return %[[VAL_7]] : memref<?xi32, 4>
 // CHECK-NEXT:    }
```
Comment (victor-eds) [inline mlir-sycl/test/Transforms/sycl-method-to-sycl-call.mlir:25]: We changed the argument type here.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 10]
```diff
@@ -1058,28 +1059,6 @@ static NamedAttrList getSYCLMethodOpAttrs(OpBuilder &Builder,
...
-          [](Operation *Op) {
-            if (auto *Res = trackCasts(Op->getOperand(0)))
-              return Res;
-            return Op;
-          })
-      .Case<mlir::polygeist::Pointer2MemrefOp, mlir::LLVM::AddrSpaceCastOp>(
-          [](Operation *Op) { return trackCasts(Op->getOperand(0)); })
-      .Default(static_cast<Operation *>(nullptr));
-}
-
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1082]: Moved.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 11]
```diff
@@ -1118,15 +1092,12 @@ llvm::Optional<sycl::SYCLMethodOpInterface> MLIRScanner::createSYCLMethodOp(
   LLVM_DEBUG(llvm::dbgs() << "Inserting operation " << OptOpName
                           << " to replace SYCL method call.\n");
 
-  OperandsCpy[0] = Builder.createOrFold<memref::LoadOp>(
-      Loc, OperandsCpy[0],
-      Builder.createOrFold<arith::ConstantIndexOp>(Loc, 0));
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1123]: Replaced by `sycl::adaptSYCLMethodOpArguments` call.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**